lib/web_ui/lib/src/engine/text/paragraph.dart

```diff
@@ -4,6 +4,107 @@
 
 part of engine;
 
+class EngineLineMetrics implements ui.LineMetrics {
+  EngineLineMetrics({
+    this.hardBreak,
+    this.ascent,
+    this.descent,
+    this.unscaledAscent,
+    this.height,
+    this.width,
+    this.left,
+    this.baseline,
+    this.lineNumber,
+  }) : text = null;
+
+  EngineLineMetrics.withText(
+    this.text, {
+    @required this.hardBreak,
+    this.ascent,
+    this.descent,
+    this.unscaledAscent,
+    this.height,
+    @required this.width,
+    this.left,
+    this.baseline,
+    @required this.lineNumber,
+  })  : assert(text != null),
+        assert(hardBreak != null),
+        assert(width != null),
+        assert(lineNumber != null && lineNumber >= 0);
+
+  /// The textual content representing this line.
+  final String text;
+
+  @override
+  final bool hardBreak;
+
+  @override
+  final double ascent;
+
+  @override
+  final double descent;
+
+  @override
+  final double unscaledAscent;
+
+  @override
+  final double height;
+
+  @override
+  final double width;
+
+  @override
+  final double left;
+
+  @override
+  final double baseline;
+
+  @override
+  final int lineNumber;
+
+  @override
+  int get hashCode => ui.hashValues(
+        text,
+        hardBreak,
+        ascent,
+        descent,
+        unscaledAscent,
+        height,
+        width,
+        left,
+        baseline,
+        lineNumber,
+      );
+
+  @override
+  bool operator ==(dynamic other) {
+    if (identical(this, other)) {
+      return true;
+    }
+
+    if (other.runtimeType != runtimeType) {
+      return false;
+    }
+    final EngineLineMetrics typedOther = other;
+    return text == typedOther.text &&
+        hardBreak == typedOther.hardBreak &&
+        ascent == typedOther.ascent &&
+        descent == typedOther.descent &&
+        unscaledAscent == typedOther.unscaledAscent &&
+        height == typedOther.height &&
+        width == typedOther.width &&
+        left == typedOther.left &&
+        baseline == typedOther.baseline &&
+        lineNumber == typedOther.lineNumber;
+  }
+
+  @override
+  String toString() {
+    return 'EngineLineMetrics(lineNumber: $lineNumber, text: "$text", hardBreak: $hardBreak, width: $width)';
```

Comment:
This string could show up in user code (there was a recent test breakage when some web class went out of sync). Let's make sure we're consistent with the C++ version.

Reply:
This is only useful for tests now, so I'll hide it behind `assertionsEnabled`. The C++ version doesn't implement it. It just return the default `Instance of 'LineMetrics'`.

Reply:
Maybe add the same implementation in the C++ version?

Reply:
Removed the method.